### ISSUE TITLE
Intermittent CI failures in sslapitest

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#23677]
The CI failures are happenning in memory_sanitizer or ub_sanitizer builds.
Example:

```
# INFO:  @ test/helpers/ssltestlib.c:1280
        # SSL_accept() failed -1, 1
        # 000822B58E7F0000:error:0A000065:SSL routines:final_key_share:no suitable key share:ssl/statem/extensions.c:1470:
        # INFO:  @ test/helpers/ssltestlib.c:1246
        # SSL_connect() failed -1, 1
        # 000822B58E7F0000:error:0A000410:SSL routines:ssl3_read_bytes:ssl/tls alert handshake failure:ssl/record/rec_layer_s3.c:907:SSL alert number 40
        # ERROR: (bool) 'create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE) == true' failed @ test/sslapitest.c:9798
        # false
        # OPENSSL_TEST_RAND_ORDER=1708629054
        not ok [310](https://github.com/openssl/openssl/actions/runs/8007377437/job/21878781626?pr=23663#step:7:311) - iteration 3
# ------------------------------------------------------------------------------
    # OPENSSL_TEST_RAND_ORDER=1708629054
    not ok 73 - test_pluggable_signature
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/sslapitest ../../test/certs ../../test/recipes/90-test_sslapi_data/passwd.txt /tmp/Q0Dc24tufq default ../../test/default.cnf ../../test/recipes/90-test_sslapi_data/dhparams.pem => 1
not ok 1 - running sslapitest
# ------------------------------------------------------------------------------
# Looks like you failed 1 test of 4.90-test_sslapi.t ........................ 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/4 subtests
```

**Comment by t8m** [comment]
A little bit different intermittent failure happened here:
https://github.com/openssl/openssl/actions/runs/8006654162/job/21868913937

again in sslapitest but in test_early_data_skip_abort

**Comment by nhorman** [comment]
I've seen this as well.  Unfortunately theres no reported ub or memory error associated with it, and it seems to not reproduce locally here with clang 17.

**Comment by bernd-edlinger** [comment]
I suspect that #23625 introduced this.
this code overwrites xor_group.group_id:
https://github.com/openssl/openssl/blob/2d70cc9cecf8b322d795985efecee06242b203b3/test/tls-provider.c#L413-L414
but the wrong value is used when this function is called a second time:
https://github.com/openssl/openssl/blob/2d70cc9cecf8b322d795985efecee06242b203b3/test/tls-provider.c#L388
and as a consequence this assertion does not hold:
```
index e5737dc..9cbd76d 100644
--- a/test/tls-provider.c
+++ b/test/tls-provider.c
@@ -385,6 +385,7 @@ static int tls_prov_get_capabilities(void *provctx, const char *capability,
 
     if (strcmp(capability, "TLS-GROUP") == 0) {
         /* Register our 2 groups */
+        OPENSSL_assert(xor_group.group_id >= 65024 && xor_group.group_id < 65279 - NUM_DUMMY_GROUPS);
         ret = cb(xor_group_params, arg);
         ret &= cb(xor_kemgroup_params, arg);
 
```

**Comment by t8m** [comment]
This is fixed by #23705
